nixos/modules/services/continuous-integration/gitlab-runner.nix

```diff
@@ -1,160 +1,454 @@
 { config, lib, pkgs, ... }:
-
 with lib;
-
 let
   cfg = config.services.gitlab-runner;
-  configFile =
-    if (cfg.configFile == null) then
-      (pkgs.runCommand "config.toml" {
-        buildInputs = [ pkgs.remarshal ];
-        preferLocalBuild = true;
-      } ''
-        remarshal -if json -of toml \
-          < ${pkgs.writeText "config.json" (builtins.toJSON cfg.configOptions)} \
-          > $out
-      '')
-    else
-      cfg.configFile;
   hasDocker = config.virtualisation.docker.enable;
+  hashedServices = with builtins; (mapAttrs' (name: service: nameValuePair
+    "${name}_${config.networking.hostName}_${unsafeDiscardStringContext (
+        substring 0 12
+        (hashString "md5" (toJSON service)))}"
+      service)
+    cfg.services);
+  configPath = "/var/lib/gitlab-runner/.gitlab-runner/config.toml";
+  configureScript = pkgs.writeShellScriptBin "gitlab-runner-configure" (
+    if (cfg.configFile != null) then ''
+      mkdir -p $(dirname ${configPath})
+      cp ${cfg.configFile} ${configPath}
```

Comment:
I did another rebuild today and this line broke witht he following error:

```
gitlab-runner-configure[8328]: cp: cannot create regular file '/var/lib/gitlab-runner/.gitlab-runner/config.toml': Permission denied
```

The reason I think is that the uid of `/var/lib/private/gitlab-runner`, which is the directory `/var/lib/gitlab-runner` points to, has a different uid. I asume that has to do with the usage of `DynamicUser`.

Are you sure this always works in the case of a generated config file?

Should it really be an issue with `DynamicUser` I'm wondering if it is worht using it. The uid/gid for gitlab-runner are already reserved in NixOS so we might as well use them. It might also make things easier when the shell executor is used as the uids are predictable.

Reply:
I had no issues with generated config file. Done a lot of rebuilds/restarts and never had issues with permissions.
`systemd` should `chown` files in `/var/lib/gitlab-runner` on every start but check output of `sudo ls -l /var/lib/gitlab-runner/.gitlab-runner/config.toml` to make sure.
Also check permissions for file in `configFile`, maybe it's unreadabe by service.

Reply:
Configuration script will run as root now, should fix any permission issues.

Reply:
Ok but what is the advantage of using `DynamicUser` over the existing `gitlab-runner` user?

Reply:
Existing user can be removed :)
See https://github.com/NixOS/rfcs/pull/52

Reply:
@bachp you might have run into the following:

> Except in case of ConfigurationDirectory=, the innermost specified directories will be owned by the user and group
> specified in User= and Group=. If the specified directories already exist and their owning user or group do not match the
> configured ones, all files and directories below the specified directories as well as the directories themselves will have
> their file ownership recursively changed to match what is configured. As an optimization, if the specified directories are
> already owned by the right user and group, files and directories below of them are left as-is, even if they do not match
> what is requested. The innermost specified directories will have their access mode adjusted to the what is specified in
> RuntimeDirectoryMode=, StateDirectoryMode=, CacheDirectoryMode=, LogsDirectoryMode= and ConfigurationDirectoryMode=.

I can't think of many reasons about how this could have happened, except leftover state from a previous version of the module. I'd prefer if we could keep not running the configuration script as root. If the reason @bachp ran into is a valid problem during migration, we could set
```
StateDirectory = "gitlab-runner gitlab-runner/.gitlab-runner";
```

to ensure files in there are properly chown'ed - I assume we want to keep the `.gitlab-runner` folder, and not just place `config.toml` at `/var/lib/gitlab-runner/config.toml`, right?

Reply:
> I'd prefer if we could keep not running the configuration script as root.

Configuration files for service will have to be readable by all users or be under `/var/lib/gitlab-runner` directory since you can't just `chown` for dynamically created user.
Running configuration script as `root` allows reading any configuration files.

Reply:
If you set `StateDirectory` as described above, systemd takes care of doing the `chown` before running the script as the dynamic user.

Reply:
> If you set `StateDirectory` as described above, systemd takes care of doing the `chown` before running the script as the dynamic user.

This seems to break everything:
```
апр 14 18:08:20 localhost systemd[13466]: gitlab-runner.service: Failed to set up special execution directory in /var/lib: File exists
апр 14 18:08:20 localhost systemd[13466]: gitlab-runner.service: Failed at step STATE_DIRECTORY spawning /nix/store/5i32p4sygpy1lxcfjajyfmsil6174g7c-gitlab-runner-configure/bin/gitlab-runner-configure: File exists
```
Removing `/var/lib/gitlab-runner` and `/var/lib/private/gitlab-runner` doesn't help.